ruby/type-hinting-converters-codecs-errors.md

```diff
@@ -0,0 +1,273 @@
+  end
+end
+
+class MyWorkflow < Temporal::Workflow
+  input MyInput
+  output MyOutput
+
+  def execute(input)
+    MyOutput.new(input.a + input.b)
+  end
+end
+```
+
+Custom types are expected to implement the following interface:
+
+```ruby
+interface _CustomType[U]
+  # Take a primitive type and return an instance of the implementing class
+  def self.load: (U) -> instance
+
+  # Returns a trivially serializable type representing the custom object
+  def dump: -> U
+end
+```
+
+The same exact concept applies to signals and queries:
+
+```ruby
+class MyWorkflow < Temporal::Workflow
+  signal :add, input: Integer
+  query :get_sum, output: MyOutput
+
+  def execute
+    ...
+  end
+end
+```
+
+Notes:
+
+- This feature is 100% optional. Without these definitions the output from converters will be used
+  directly as arguments and return values
+- The full potential of this solution is obviously limited to a single language SDKs
+- When sending a custom input type to another SDK, workflow/activity definitions (without the
+  implementation) can still be leveraged
+- A custom output type sent to another SDK can still be `dump`ed into a matching JSON or other
+  serializable structure
+- The same approaches apply to inputs and outputs received from other SDKs
+
+
+## Payload handling
+
+Ruby SDK will use the model proposed [here](https://docs.temporal.io/docs/concepts/what-is-a-data-converter/)
+to distinguish between payload converters and codecs.
+
+The main difference is that a Converter translates between any Ruby object and a Payload, while
+a Codec is a lower level construct that encodes/decodes from a Payloads protobuf to a Payloads
+protobuf. The former is intended to support specific serialisation formats, while the latter is for
+modifying a binary that is transmitted over the wire.
+
+
+### Converters
+
+A payload converter is expected to implement the following interface:
+
+```ruby
+interface _PayloadConverter[U]
+  # Returns whether or not it can decode a given Payload
+  def can_decode?: (Temporal::Payload) -> Bool
+
+  # Takes any Ruby object and returns a Payload object or nil if unable to convert
+  def to_payload: (U) -> Temporal::Payload
+
+  # Takes a Payload object and returns a Ruby object. Only called on a matching encoding
+  def from_payload: (Temporal::Payload) -> U
+end
+```
+
+The SDK will provide an implementation of basic converters — Null (for `nil`s), Bytes (any stream of
+bytes) and JSON (using the `oj` gem and potentially allowing the SDK user to specify the
+[mode](https://github.com/ohler55/oj/blob/develop/pages/Modes.md)). We will also provide a
+`Composite` converter to combine multiple converters behind the same interface.
+
+
+### Codecs
+
+And a payload codec is expected to implement this interface:
+
+```ruby
+interface _PayloadCodec
+  # Takes an array of Payload objects and returns an array of encoded Payload object
+  def encode: (Array[Temporal::Payload]) -> Array[Temporal::Payload]
```

Comment:
Need to clarify that while the resulting array doesn't have to be the same length as the parameter array, it does have to have at least 1 value. Same for `decode`.

Reply:
Added the clarification